### PR TITLE
Update oversampling module with new api, examples and doc

### PR DESCRIPTION
**Pull request by Anaxagor**
## BNOversampling module
This pull request contains changes to the synthetic generation module and in particular adds an oversampling module that solves the problem of class imbalance in the data.

This pull request includes significant updates to the `applybn` package, focusing on improving the Bayesian Network-based oversampling functionality, adding new features, and enhancing documentation and tests. The most important changes include the addition of a new oversampling class, updates to the synthetic data generator, and new utility functions for Bayesian Networks.

### Enhancements to Bayesian Network-based Oversampling:

* **New Oversampling Class:**
  * [`applybn/imbalanced/over_sampling/bn_over_sampler.py`](diffhunk://#diff-5d0df5870df9df4537974de1626e0f4bfde606b87214cd906ddcc0a917db5a5fL2-R142): Introduced the `BNOverSampler` class, which uses Bayesian Networks to generate synthetic samples for imbalanced datasets. The class includes parameters for specifying the target class column, oversampling strategy, and shuffle option.

* **Synthetic Data Generator:**
  * [`applybn/synthetics/bn_synthetic_generator.py`](diffhunk://#diff-55ece40ed01c9c3390bc0f6e64b0e76d4aa23a87c762a3785ea8e535b3ceffabL1-R14): Refactored the `BNSyntheticGenerator` class to include methods for fitting a Bayesian Network model to data and generating synthetic samples.

* **Utility Function for Bayesian Networks:**
  * [`applybn/utils/bn_wrapper.py`](diffhunk://#diff-bc313b92872186755d2ae111d6bde5f6f2b70787d950681b0e84b69abc5457b9R1-R52): Added the `get_bn` function to preprocess data and build a Bayesian Network, either hybrid or discrete, based on the data types.

### Documentation and Examples:

* **API Documentation:**
  * [`docs/api/bn_oversampling.md`](diffhunk://#diff-4060ad33282bcd25b71383d2cb6ebd11ca20d1ad33749c012f31378f689aa266R4-R9): Updated the documentation for the `BNOverSampler` class, including an example usage section.

* **Example Script:**
  * [`examples/imbalanced/iris_example.py`](diffhunk://#diff-f32966481fd7dbb7bb279ad89b6a08b52112c045249f0a6170d40afc5ef1d173R1-R94): Added an example script demonstrating the use of `BNOverSampler` with the Iris dataset, including creating an imbalanced dataset, resampling, and logging.

### Testing and Dependencies:

* **New Test Fixtures and Cases:**
  * [`tests/conftest.py`](diffhunk://#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R1-R22): Added fixtures for generating imbalanced and mixed-type datasets.
  * [`tests/test_bn_generator.py`](diffhunk://#diff-d72cf9bc5925707e021e2eb8c44e925f48d89e409216c9971528862b9b0d60d9R1-R15): Added tests for the `BNSyntheticGenerator` class, including initialization and fitting.
  * [`tests/test_bn_over_sampler.py`](diffhunk://#diff-914ee4283b808551f93ec5e4133fb289b4edf5495fdf0a220b837364eec1a3edR1-R52): Added tests for the `BNOverSampler` class, including initialization, basic oversampling, custom strategies, and mixed data types.

* **Dependency Updates:**
  * [`pyproject.toml`](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711L12-R19): Updated dependencies to ensure compatibility, including changes to `numpy`, `pandas`, and adding new dependencies such as `bamt` and `setuptools`.

These changes collectively enhance the functionality, usability, and reliability of the `applybn` package for handling imbalanced datasets using Bayesian Networks.

### CODE REVIEW

**Comment by jrzkaminski** [comment]
A general comment:
All the docstrings should have Google style Forman without typing hints.
Typing hints should be implemented in definitions

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 1]
```diff
@@ -1,14 +1,142 @@
...
+                        samples[c] = samples[c].astype(int)
+                balanced_data = pd.concat([balanced_data, samples], ignore_index=True)
+        # shuffle data
+        if self.shuffle:
+            balanced_data = balanced_data.sample(frac=1).reset_index(drop=True)
+        # Split back into features and target
+        X_res = balanced_data.drop(columns=[self.class_column])
+        y_res = balanced_data[self.class_column]
+
+        return X_res, y_res
```
Comment (jrzkaminski) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:142]: Probably that needs decomposing.

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 2]
```diff
@@ -1,14 +1,142 @@
...
+                if samples.shape[0] < needed:
+                    additional_samples = self.data_generator_.bn.sample(
+                    needed, 
+                    evidence={self.class_column: cls}, filter_neg=False
+                )[data.columns]
+                samples = pd.concat([samples, additional_samples.sample(needed - samples.shape[0])])
+                for c in samples.columns:
+                    if types_dict[c] == 'disc_num':
+                        samples[c] = samples[c].astype(int)
+                balanced_data = pd.concat([balanced_data, samples], ignore_index=True)
```
Comment (jrzkaminski) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:134]: anything we can do about this nested loop? Use pd.apply at least?

**Review by jrzkaminski — CHANGES_REQUESTED** [review]

File: applybn/utils/bn_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        >>> data = pd.DataFrame({"A": [1, 2, 3], "B": ["X", "Y", "X"]})
+        >>> bn = get_bn(data, n_bins=5, has_mixture=False)
+        >>> bn.type  # Returns 'Discrete' based on data types
+
+    Notes:
+        - Preprocessing: Combines LabelEncoder (for categoricals) and KBinsDiscretizer (for 
+          continuous features) via a Preprocessor.
+        - Structure Learning: Uses the K2 scoring function to determine edges.
+        - HybridBN: Used when continuous features are detected. DiscreteBN is a fallback.
+    """
```
[the remark this answers is not in the document]
Comment (Anaxagor) [reply applybn/utils/bn_wrapper.py]: Done

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 4]
```diff
@@ -1,14 +1,142 @@
...
+        NotFittedError
+            If synthetic generator fails to fit Bayesian Network
+
+        Notes
+        -----
+        1. Combines X and y into single DataFrame for Bayesian Network learning
+        2. Determines target sample sizes based on strategy
+        3. Generates synthetic samples for minority classes using conditional sampling
+        4. Preserves original data types and column names
+        """
```
[the remark this answers is not in the document]
Comment (Anaxagor) [reply applybn/imbalanced/over_sampling/bn_over_sampler.py]: Done

File: examples/imbalanced/iris_example.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    
+    Args:
+        X (pd.DataFrame): Original feature matrix
+        y (pd.Series): Original target vector
+        ratios (dict): Dictionary mapping class labels to desired sample counts
+        
+    Returns:
+        tuple: (imbalanced_X, imbalanced_y) with specified class distribution
+ 
+    """
```
[the remark this answers is not in the document]
Comment (Anaxagor) [reply examples/imbalanced/iris_example.py]: Done

File: applybn/imbalanced/over_sampling/bn_over_sampler.py (hunk 1 above)
Comment (Anaxagor) [reply applybn/imbalanced/over_sampling/bn_over_sampler.py]: Done

File: applybn/imbalanced/over_sampling/bn_over_sampler.py (hunk 2 above)
Comment (Anaxagor) [reply applybn/imbalanced/over_sampling/bn_over_sampler.py]: Done

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 6]
```diff
@@ -1,14 +1,152 @@
...
+            balanced_data: Balanced dataset containing original and synthetic samples.
+        """
+        balanced_data = data.copy()
+        types_dict = self.data_generator_.bn.descriptor['types']
+        for cls in class_counts.index:
+            needed = max(0, target_size - class_counts[cls])
+            if needed > 0:
+                samples = self._generate_samples_for_class(cls, needed, data.columns, types_dict)
+                balanced_data = pd.concat([balanced_data, samples], ignore_index=True)
+        return balanced_data
```
Comment (jrzkaminski) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:100]:
```suggestion
        samples = []
        types_dict = self.data_generator_.bn.descriptor['types']
        
        # Calculate needed samples for each class
        needed_samples = (target_size - class_counts).clip(lower=0)
        
        # Generate samples for classes requiring augmentation
        for cls, needed in needed_samples.items():
            if needed > 0:
                samples.append(self._generate_samples_for_class(cls, needed, data.columns, types_dict))
        
        # Combine original data with all generated samples at once
        return pd.concat([data] + samples, ignore_index=True) if samples else data.copy()
```

File: applybn/utils/bn_wrapper.py
Comment (jrzkaminski) [inline applybn/utils/bn_wrapper.py:1]: Are we waiting for [this PR](https://github.com/Anaxagor/applybn/pull/38) to be merged to replace this utility?

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 7]
```diff
@@ -1,14 +1,167 @@
...
+        y_series = pd.Series(y) if not isinstance(y, pd.Series) else y.copy()
+        if self.class_column is None:
+            self.class_column = y.name if hasattr(y, 'name') else 'class'
+        data = X_df.assign(**{self.class_column: y_series})
+        
+        # Preprocess data
+        encoder = LabelEncoder()
+        discretizer = KBinsDiscretizer(n_bins=10, encode='ordinal', strategy='quantile')
+        pp = Preprocessor([("encoder", encoder), ("discretizer", discretizer)])
+        preprocessed_data, _ = pp.apply(data)
```
Comment (Roman223) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:145]: The default preprocessor can be taken from core

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 8]
```diff
@@ -1,14 +1,167 @@
...
+        # Fit Bayesian Network
+        self.data_generator_.use_mixture = True
```
Comment (Roman223) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:147]: Unsafe setting attribute. Not an error, but by doing so you bypass sklearn's params validation

File: applybn/imbalanced/over_sampling/bn_over_sampler.py [hunk 9]
```diff
@@ -1,14 +1,167 @@
...
+            cls: Target class value to generate samples for.
+            needed: Number of synthetic samples needed for this class.
+            data_columns: List of column names in the original dataset.
+            types_dict: Dictionary mapping columns to their data types 
+                (e.g., 'disc_num' for discrete numeric).
+
+        Returns:
+            samples: Generated samples with proper data types.
+        """
+        samples = self.data_generator_.bn_.sample(
```
Comment (Roman223) [inline applybn/imbalanced/over_sampling/bn_over_sampler.py:56]: Delegating is not working?..
Can u just take `data_generator.bn_`?

File: tests/test_bn_over_sampler.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    assert class_counts[1] == 90
+
+def test_custom_strategy(imbalanced_data):
+    sampler = BNOverSampler(strategy=150, class_column='class')
+    X = imbalanced_data.drop(columns='class')
+    y = imbalanced_data['class']
+    
+    X_res, y_res = sampler._fit_resample(X, y)
+    
+    # Check total samples (original 100 + synthetic 140)
```
Comment (jrzkaminski) [inline tests/test_bn_over_sampler.py:37]: 100+140 = 300?

File: examples/imbalanced/iris_example.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    oversampler = BNOverSampler(
+        class_column='species',
+        strategy='max_class',
+        shuffle=True
+    )
+    logger.debug("Initialized BNOverSampler with parameters: %s", vars(oversampler))
+
+    # Perform resampling
+    try:
+        X_res, y_res = oversampler._fit_resample(imbalanced_X, imbalanced_y)
```
Comment (jrzkaminski) [inline examples/imbalanced/iris_example.py:81]: https://github.com/scikit-learn-contrib/imbalanced-learn/blob/8189ba7fc1d936de244e120389802654b87b9c29/imblearn/base.py#L181

```suggestion
        X_res, y_res = oversampler.fit_resample(imbalanced_X, imbalanced_y)
```

File: tests/test_bn_over_sampler.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    class_counts = y_res.value_counts()
+    assert class_counts[0] == 90
+    assert class_counts[1] == 90
+
+def test_custom_strategy(imbalanced_data):
+    sampler = BNOverSampler(strategy=150, class_column='class')
+    X = imbalanced_data.drop(columns='class')
+    y = imbalanced_data['class']
+    
+    X_res, y_res = sampler._fit_resample(X, y)
```
Comment (jrzkaminski) [inline tests/test_bn_over_sampler.py:35]: Here too
```suggestion
    X_res, y_res = sampler.fit_resample(X, y)
```

File: applybn/imbalanced/over_sampling/bn_over_sampler.py (hunk 8 above)
Comment (jrzkaminski) [reply applybn/imbalanced/over_sampling/bn_over_sampler.py:147]: +1
Better use @property? @Roman223

File: applybn/imbalanced/over_sampling/bn_over_sampler.py (hunk 7 above)
Comment (jrzkaminski) [reply applybn/imbalanced/over_sampling/bn_over_sampler.py:145]: where is it located?

**Review by jrzkaminski — APPROVED** [review]
